tests/src/junit/java/net/neoforged/neoforge/unittest/transfer/ItemTests.java

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.unittest.transfer;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import net.minecraft.core.Direction;
+import net.minecraft.core.component.DataComponents;
+import net.minecraft.world.Container;
+import net.minecraft.world.SimpleContainer;
+import net.minecraft.world.WorldlyContainer;
+import net.minecraft.world.inventory.AbstractContainerMenu;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
+import net.neoforged.neoforge.transfer.ResourceHandlerUtil;
+import net.neoforged.neoforge.transfer.handlers.wrappers.items.VanillaContainerWrapper;
+import net.neoforged.neoforge.transfer.handlers.wrappers.items.WorldlyContainerWrapper;
+import net.neoforged.neoforge.transfer.resources.ItemResource;
+import net.neoforged.neoforge.transfer.transaction.Transaction;
+import org.jetbrains.annotations.Nullable;
+import org.junit.jupiter.api.Test;
+
+public class ItemTests {
+    @Test
+    public void testStackReference() {
+        // Ensure that Container wrappers will try to mutate the backing stack as much as possible.
+        // In many cases, MC code captures a reference to the ItemStack so we want to edit that stack directly
+        // and not a copy whenever we can. Obviously this can't be perfect, but we try to cover as many cases as possible.
+        SimpleContainer container = new SimpleContainer(new ItemStack(Items.DIAMOND, 2));
+        var containerWrapper = VanillaContainerWrapper.of(container);
+        ItemStack stack = container.getItem(0);
+
+        // Simulate should correctly reset the stack.
+        try (Transaction tx = Transaction.open(null)) {
+            containerWrapper.extract(ItemResource.of(Items.DIAMOND), 2, tx);
+        }
+
+        if (stack != container.getItem(0)) throw new AssertionError("Stack should have stayed the same.");
+
+        // Commit should try to edit the original stack when it is feasible to do so.
+        try (Transaction tx = Transaction.open(null)) {
+            containerWrapper.extract(ItemResource.of(Items.DIAMOND), 1, tx);
+            tx.commit();
+        }
+
+        if (stack != container.getItem(0)) throw new AssertionError("Stack should have stayed the same.");
+
+        // Also edit the stack when the item matches, even when the components and the count change.
+        ItemResource oldResource = ItemResource.of(Items.DIAMOND);
+        ItemResource newResource = oldResource.with(DataComponents.MAX_DAMAGE, 10);
+
+        try (Transaction tx = Transaction.open(null)) {
+            containerWrapper.extract(oldResource, 2, tx);
```

Comment:
Do we want to be using `2` or `1` here? Given theoretically the stack's size at this point is one, yes extracting with more than the amount should be fine, but `2` feels like a semi arbitrary number to do so with, rather than say the resource's max stack size

Reply:
1 makes the most sense to me indeed. Changed.